src/java/com/rapleaf/jack/queries/GenericQuery.java

```
@@ -0,0 +1,147 @@
+  private final List<OrderCriterion> orderCriteria;
+  private final Set<ModelField> selectedModelFields;
+  private Optional<LimitCriterion> limitCriteria;
+
+  private GenericQuery(BaseDatabaseConnection dbConnection) {
+    this.dbConnection = dbConnection;
+    this.mainModel = null;
+    this.joinConditions = Lists.newArrayList();
+    this.whereConstraints = Lists.newArrayList();
+    this.orderCriteria = Lists.newArrayList();
+    this.selectedModelFields = Sets.newHashSet();
+    this.limitCriteria = Optional.absent();
+  }
+
+  public static GenericQuery create(BaseDatabaseConnection dbConnection) {
+    return new GenericQuery(dbConnection);
+  }
+
+  public GenericQueryBuilder from(Class<? extends ModelWithId> model) {
+    this.mainModel = model;
+    return new GenericQueryBuilder(dbConnection, this);
+  }
+
+  List<WhereConstraint> getWhereConstraints() {
+    return whereConstraints;
+  }
+
+  Set<ModelField> getSelectedModelFields() {
+    return selectedModelFields;
+  }
+
+  void addJoinCondition(JoinCondition joinCondition) {
+    this.joinConditions.add(joinCondition);
+  }
+
+  void addWhereCondition(WhereConstraint whereConstraint) {
+    this.whereConstraints.add(whereConstraint);
+  }
+
+  void addOrderCondition(OrderCriterion orderCriterion) {
+    this.orderCriteria.add(orderCriterion);
+  }
+
+  void addLimitCondition(LimitCriterion limitCriterion) {
+    this.limitCriteria = Optional.of(limitCriterion);
+  }
+
+  void addSelectedModelField(ModelField modelField) {
+    this.selectedModelFields.add(modelField);
+  }
+
+  boolean isOrderedQuery() {
+    return !orderCriteria.isEmpty();
+  }
+
+  String getSqlStatement() {
+    StringBuilder statement = new StringBuilder();
+
+    statement.append(getSelectClause())
+        .append(getJoinClause())
+        .append(getWhereClause());
+
+    if (isOrderedQuery()) {
+      statement.append(getOrderClause());
+      statement.append(getLimitClause());
```

Do you want to chat about this offline? I feel somewhat strongly that we should support limit without order. Also, if we didn't we should raise an exception if there was a limit so people know their input is being ignored.